File: benchmark/sort.c
#include <stdio.h>
#include <time.h>

#define N 10000
#define ITERS 50

static int lcg(int state) {
    return (state * 1103 + 12345) % 1000000;
}

static void swap(int *a, int *b) {
    int t = *a; *a = *b; *b = t;
}

static void quicksort(int *a, int lo, int hi) {
    if (lo < hi) {
        int pivot_val = a[hi];
        int i = lo;
        for (int j = lo; j < hi; j++) {
            if (a[j] < pivot_val) {
                swap(&a[i], &a[j]);
                i++;
            }
        }
        swap(&a[i], &a[hi]);
        quicksort(a, lo, i - 1);
        quicksort(a, i + 1, hi);
    }
}

int main(void) {
    int a[N];
    int checksum = 0;

    clock_t t0 = clock();

    for (int iter = 0; iter < ITERS; iter++) {
        int seed = iter * 7 + 42;
        for (int i = 0; i < N; i++) {
            seed = lcg(seed);
            a[i] = seed;
        }
        quicksort(a, 0, N - 1);
        checksum += a[0] + a[N - 1];
    }

    double elapsed = (double)(clock() - t0) / CLOCKS_PER_SEC;
    fprintf(stderr, "exec: %.3fs\n", elapsed);

    volatile int sink = checksum;
    (void)sink;
    return 0;
}

File: benchmark/sort.lua
-- Sorting benchmark
-- Quicksort 10000 elements, 50 iterations

local N = 10000
local ITERS = 50

local function lcg(state)
    return (state * 1103 + 12345) % 1000000
end

local function quicksort(a, lo, hi)
    if lo < hi then
        local pivot_val = a[hi]
        local i = lo
        for j = lo, hi - 1 do
            if a[j] < pivot_val then
                a[i], a[j] = a[j], a[i]
                i = i + 1
            end
        end
        a[i], a[hi] = a[hi], a[i]
        quicksort(a, lo, i - 1)
        quicksort(a, i + 1, hi)
    end
end

local function run()
    local a = {}
    local checksum = 0

    for iter = 0, ITERS - 1 do
        local seed = iter * 7 + 42
        for i = 1, N do
            seed = lcg(seed)
            a[i] = seed
        end
        quicksort(a, 1, N)
        checksum = checksum + a[1] + a[N]
    end
end

local t0 = os.clock()
run()
local elapsed = os.clock() - t0
io.stderr:write(string.format("exec: %.3fs\n", elapsed))
